daemon/indexing/apollo/index.js

```diff
@@ -4,37 +4,170 @@ var search = require('../lib/search.js')
 var db = require('../lib/db.js')
 
 /*
- * Implentation of a GraphQL server using the Apollo framework.
- * See https://www.apollographql.com/server
+ * Implementation of the Origin GraphQL server.
+  * Uses the Apollo framework: https://www.apollographql.com/server
  *
  */
 
 // Type definitions define the "shape" of the data and specify
 // which ways the data can be fetched from the GraphQL server.
 const typeDefs = gql`
+  ######################
+  #
+  # Query output schema.
+  #
+  ######################
+
+  type User {
+    walletAddr: ID!   # Ethereum wallet address
+    identityAddr: ID  # ERC 725 identity address.
+  }
+
+  type Offer {
+    id: ID!
+    ipfsHash: ID!
+    listingId: ID!
+    buyer: User!
+    # Per contract definition: 1: Created, 2: Accepted, 3: Disputed
+    # NOTE: There is no "Finalized" status stored on-chain but could be useful
+    #       to compute it when resolving this object.
+    status: Int!
+  }
+
+  type Price {
+    currency: String!
+    amount: Float!
+  }
+
+  type Review {
+    ipfsHash: ID!
+    reviewer: User!
+    text: String!
+    rating: Int!
+  }
+
+  # TODO: Add a status indicating if Listing is sold out.
   type Listing {
-    id: String
-    name: String
+    id: ID!
+    ipfsHash: ID!
+    seller: User!
+    title: String!
     description: String
-    price: Float
+    price: Price!
+    offers: [Offer]
+    reviews: [Review]
+  }
+
+  ######################
+  #
+  # Query input schema.
+  #
+  ######################
+
+  input Page {
+    num: Int!  # Page number.
+    size: Int! # Number of items per page.
+  }
+
+  # Note: Defined as inPrice vs Price because GraphQL does not allow to
+  # use same name for input and output types.
+  input inPrice {
+    currency: String!
+    amount: Float!
+  }
+
+  # TODO:
+  #  - Filtering definition needs more thinking. This is not flexible at all...
+  #  - Add location based filtering.
+  #  - Add fractional usage (e.g. availability) filtering.
+  input ListingFilter {
+    priceMin: inPrice
+    priceMax: inPrice
+    cat: String
+    subCat: String
+    locale: String
+    sellerAddr: String
+    buyerAddr: String
+  }
+
+  enum ListingOrderBy {
+    relevance         # Default if no order by specified.
+    priceAsc          # Price low to high.
+    priceDesc         # Price high to low.
+    creationDateDesc  # Most to least recent.
+    sellerRating      # Highest to lowest rating.
+  }
+
+  input ListingQuery {
+    search: String    # Search query. If not specified, all listings are candidates.
+    filter: ListingFilter
+    orderBy: ListingOrderBy
+    page: Page
   }
 
   # The "Query" type is the root of all GraphQL queries.
   type Query {
-    Listings(query: String): [Listing],
+    Listings(query: ListingQuery!): [Listing],
+    Listing(id: ID!): Listing,
```

Comment:
We need to allow querying down from Users and Offers as well.

A common use case would be to get a list of all my active offers. (Forgive the half thought out code below.)

```
Users(id:"me"){
  offers(active:True){
      id,
      ...,
      listing{
        id,
        name
      }
  }
}
```

Reply:
If we are skipping the nodes and edges stuff, we still need a way to return the number of results for a paginated listing.

For example, if someone has 1,000 listings for sale, we need to be able to get listings between 100 and 150 listings, while also showing the total number of listings so that we can show the appropriate navigation options on the client.